x-pack/plugins/security_solution/public/onboarding/types.ts

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type React from 'react';
+import type { IconType } from '@elastic/eui';
+import type { LicenseType } from '@kbn/licensing-plugin/public';
+
+import type { OnboardingCardId } from './constants';
+import type { RequiredCapabilities } from '../common/lib/capabilities';
+import type { StartServices } from '../types';
+
+export interface CheckCompleteResult {
+  /**
+   * Optional custom badge text replacement for the card complete badge in the card header.
+   */
+  completeBadgeText?: string | React.ReactNode;
+  /**
+   * Optional badges to prepend to the card complete badge in the card header, regardless of completion status.
+   */
+  additionalBadges?: React.ReactNode[];
+  /**
+   * Optional metadata to be passed to the card component.
+   */
+  metadata?: Record<string, unknown>;
+}
+/**
+ * The result of a card completion auto check.
+ * - `true` if the card is complete.
+ * - `false` if the card is not complete.
+ * - `{ isComplete: true, completeBadgeText: ReactNode }` if the card is complete and has a custom complete badge text.
+ * - `{ isComplete: false, additionalBadges: ReactNode[] }` if the card is complete and has to show additional badges on the header.
+ * - `{ isComplete: false, metadata: {showWarningCallOut: true} }` if the card is not complete and passes some metadata to the card component.
+ */
+export type CheckCompleteResponse = boolean | ({ isComplete: boolean } & CheckCompleteResult);
```

Comment:
not for now but just to keep in mind:

the description is comment is great but typescript allows to put it in action instead of hiding it here. With a union type or even better a discriminated union type we can carry this knowledge within the type and subtypes. This allows to get rid of significant amount of ifs and optional checks over and over again wherever this structure is used. The narrower we have the types and operate with them the cleaner is the code.

We can discuss this later if you want. Feel free to resolve this.